marketplace/deployer_util/validate.py

```diff
@@ -0,0 +1,185 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import json
+import yaml
+
+from argparse import ArgumentParser
+from bash_util import Command
+from dict_util import deep_get
+from yaml_util import load_resources_yaml
+from yaml_util import load_yaml
+
+DEPLOY_INFO_ANNOTATION = "marketplace.cloud.google.com/deploy-info"
+GOOGLE_MARKETPLACE = "x-google-marketplace"
+
+_PROG_HELP="""
+Validates and test the deployer image.
+"""
+
+class ValidationException(Exception):
+  pass
+
+def main():
+  parser = ArgumentParser(description=_PROG_HELP)
+  parser.add_argument('--metadata', help='Path to metadata file for validation. If it is not present, the validation on the metadata file will be skipped.',)
+  args = parser.parse_args()
+
+  all_resources, application = expand_and_load_resources()
+
+  metadata = None
+  if args.metadata:
+    with open(args.metadata, "r") as metadatafile:
+      metadata = json.loads(metadatafile.read())
+  
+  if metadata:  
+    validate_metadata(metadata, application)
+
+  schema = load_yaml("/data/schema.yaml")
+  validate_images(schema, all_resources)
+
+
+def validate_images(schema, resources):
+  declared_images = set([deep_get(param, "default") for key, param in schema["properties"].iteritems() 
+    if deep_get(param, GOOGLE_MARKETPLACE, "type") == "IMAGE"])
+
+  used_images = []
+  for resource in resources:
+    used_images.extend(deep_find(resource, "image"))
+  used_images = set(used_images)
+
+  undeclared_images = [image for image in used_images 
+    if image not in declared_images]
+
+  if undeclared_images:
+    error_message = "ERROR Images should be declared in schema file."
+    print(error_message)
+    for image in undeclared_images:
+      print("  " + str(image))
+
+    raise ValidationException(error_message)
+
+  print("All {} used images are declared.".format(len(used_images)))
+
+  unused_images = [image for image in declared_images 
+    if image not in used_images]
+
+  if unused_images:
+    error_message = "ERROR Images were declared but not used."
+    print(error_message)
+    for image in unused_images:
+      print("  " + str(image))
+
+    raise ValidationException(error_message)  
+
+  print("All {} declared images are used.".format(len(declared_images)))
+
+  for image in used_images:
+    print(image)
+
+  print("Images validation succeeded")
+
+
+def expand_and_load_resources():
+  """Expands the templates and charts so we have the final manifests manifests"""
+  print(Command("/bin/expand_config.py").output)
```

Comment:
A few things here:
- `expand_config` will fail if a required property doesn't have a value. For example, service account or some custom property like email address. These properties will need a test value in order to pass validation anyway, so we can use those test values.
- The test values can probably be supplied in the test `schema.yaml`, as `default` values?
- I'm thinking we should allow the partner to run this to validate their image as part of their Dockerfile. This is great because errors will be caught at build time instead of runtime. See #82.
- For this to happen, this cannot rely on modifying local file system. Thus, we should probably modify `expand_config` to make it into a library and use it here. The main output of `expand_config` is a dictionary of input parameters.
- `create_manifests.sh` is another beast..
-